### PR TITLE
Sink to MemMove optimization in injectdestructors

### PR DESCRIPTION
**Pull request by cooldome**
While working on PR  https://github.com/nim-lang/Nim/pull/12992 I came up with idea how number of sink calls be significantly reduced. `isFirstWrite` is used to detect if it safe to optimize `=sink` calls to bitwise `memmove` calls. So far  `isFirstWrite`  is true if it is `let x = ..` or `var y = ...` outside of the loop, but further PRs can potentially use control flow graph to optimize assignments as well. In similar fashion as we have `isLastRead` we also will need `isFirstWrite` to make it happen.

This PR contains all changes of https://github.com/nim-lang/Nim/pull/12992. As I used it as a starting point.

### CODE REVIEW

File: compiler/lowerings.nim [hunk 1]
```diff
@@ -60,14 +67,12 @@ proc lowerTupleUnpacking*(g: ModuleGraph; n: PNode; owner: PSym): PNode =
   var temp = newSym(skTemp, getIdent(g.cache, genPrefix), owner, value.info, g.config.options)
   temp.typ = skipTypes(value.typ, abstractInst)
   incl(temp.flags, sfFromGeneric)
-  incl(temp.flags, sfCursor)
 
   var v = newNodeI(nkVarSection, value.info)
   let tempAsNode = newSymNode(temp)
-  v.addVar(tempAsNode)
+  v.addVar(tempAsNode, value)
```
Comment (cooldome) [inline compiler/lowerings.nim]: I have changed lowerTupleUnpacking because now pattern `var x = value` is optimized while `var x; x = ...` is not.

**Comment by Araq** [comment]
This is totally part of the plan. :-)

**Comment by Clyybber** [comment]
I think its best to call these memmove calls `=weakSink`. We will need `=weakAsgn` too later on for eliding initialization.

**Comment by cooldome** [comment]
thavlak with `--gc:orc` test failures seems unrelated

**Comment by Araq** [comment]
> thavlak with --gc:orc test failures seems unrelated

How so? I spent some effort on it to ensure it's green.

**Comment by cooldome** [comment]
Apologies, I have seen this test failing before automatically thought it is unrelated.

Tests pass now, ready for review

**Comment by Araq** [comment]
It's great, you could also add a ``sfNoInit`` flags to these vars then. Also I would like to use the control flow graph for this. But I'm merging it already.

**Pull request merged**